pkg/util/log_support.go

```diff
@@ -0,0 +1,347 @@
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh/terminal"
+
+	"encoding/json"
+)
+
+var (
+	// UserOut is the customized logrus log used for direct user output
+	UserOut = log.New()
+	// UserOutFormatter is the specialized formatter for UserOut
+	UserOutFormatter = new(TextFormatter)
+	// JSONOutput is a bool telling whether we're outputting in json. Set by command-line args.
+	JSONOutput = false
+)
+
+const defaultTimestampFormat = time.RFC3339
+
+// nolint: deadcode
+const (
+	nocolor = 0
+	red     = 31
+	green   = 32 // nolint: varcheck
+	yellow  = 33
+	blue    = 36 // nolint: varcheck
+	gray    = 37
+)
+
+// TextFormatter formats logs into text.
+// This is a specialization of https://github.com/sirupsen/logrus/blob/master/text_formatter.go
+// It's intended to be used for all user-oriented output from ddev
+type TextFormatter struct {
+	// Set to true to bypass checking for a TTY before outputting colors.
+	ForceColors bool
+
+	// Force disabling colors.
+	DisableColors bool
+
+	// Disable timestamp logging. useful when output is redirected to logging
+	// system that already adds timestamps.
+	DisableTimestamp bool
+
+	// Enable logging the full timestamp when a TTY is attached instead of just
+	// the time passed since beginning of execution.
+	FullTimestamp bool
+
+	// TimestampFormat to use for display when a full timestamp is printed
+	TimestampFormat string
+
+	// The fields are sorted by default for a consistent output. For applications
+	// that log extremely frequently and don't use the JSON formatter this may not
+	// be desired.
+	DisableSorting bool
+
+	// QuoteEmptyFields will wrap empty fields in quotes if true
+	QuoteEmptyFields bool
+
+	// Whether the logger's out is to a terminal
+	isTerminal bool
+
+	sync.Once
+}
+
+// LogSetUp sets up UserOut and log loggers as needed by ddev
+func LogSetUp() {
```

Comment:
Nit: This feels more like "output" than "logs" to me. I'm not sure why. Maybe there's no difference.

Reply:
From the perspective of Logrus, it's log setup. From the perspective of the end user it's output.